### PR TITLE
Stl 1521 qa elster question

### PR DESCRIPTION
**Pull request by SannyNguyenHung**
# Short Description
- successful and maybe side differ in the detail notes. Added tests and separated the get_dependent_notes function again.

# Changes
- Delete get_dependent_notes
- Add tests for dependent_notes of the maybe page

# Feedback
- Feedback always welcome

### CODE REVIEW

File: webapp/app/model/eligibility_data.py [hunk 1]
```diff
@@ -472,10 +472,10 @@ class ForeignCountryMaybeEligibility(RecursiveDataModel):
     That's because the ForeignCountry step is the last step of the flow and needs to decide which result page is
     displayed: 'success' or 'maybe'.
     """
-    has_no_other_income: Optional[OtherIncomeEligibilityData]
     foreign_country_eligibility: str
     user_a_has_elster_account_eligibility: str
     user_b_has_elster_account_eligibility: Optional[str]
+    has_no_other_income: Optional[OtherIncomeEligibilityData]
```
Comment (frederike-ramin) [inline webapp/app/model/eligibility_data.py:478]: Why did you change this?

File: webapp/app/forms/steps/eligibility_steps.py [hunk 2]
```diff
@@ -761,23 +771,20 @@ def __init__(self, endpoint, stored_data=None, *args, **kwargs):
         
     def _main_handle(self):  
         super()._main_handle()
-              
-        self.render_info.additional_info['dependent_notes'] = get_dependent_notes(self)
+        
+        dependent_notes = []
+    
+        if data_fits_data_model_from_list([UserAElsterAccountEligibilityData, UserBElsterAccountEligibilityData], self.stored_data):
+            dependent_notes.append(_l('form.eligibility.result-note.user_b_elster_account-registration-maybe'))
```
Comment (frederike-ramin) [inline webapp/app/forms/steps/eligibility_steps.py:778]: to me, the id of the message is a bit misleading. In my head it should be something like `form.eligibility.result-note.both_elster_account-registration-maybe`

Comment (frederike-ramin) [inline webapp/app/forms/steps/eligibility_steps.py:778]: I'm not sure this is correct? The text is for the case both persons have an elster account. `data_fits_data_model_from_list` is an OR condition. So if `UserAElsterAccountEligibilityData` returns true, `UserBElsterAccountEligibilityData` is never checked. So I think only having `UserBElsterAccountEligibilityData` here would be correct as it also checks that person_a_has_elster_account_eligibility is 'yes'.

File: webapp/app/translations/de/LC_MESSAGES/messages.po [hunk 3]
```diff
@@ -1,13 +1,9 @@
-# German translations for PROJECT.
-# Copyright (C) 2020 ORGANIZATION
-# This file is distributed under the same license as the PROJECT project.
-# FIRST AUTHOR <EMAIL@ADDRESS>, 2020.
-#
+
```
Comment (frederike-ramin) [inline webapp/app/translations/de/LC_MESSAGES/messages.po:1]: Why did you delete this?

File: webapp/tests/forms/steps/test_eligibility_steps.py [hunk 4]
```diff
@@ -3167,8 +3166,8 @@ def test_if_user_b_has_no_elster_account_then_set_correct_info(self):
         self.assertEqual(expected_information, step.render_info.additional_info['dependent_notes'])
 
     def test_if_user_wants_no_cheaper_check_then_set_correct_info(self):
-        expected_information = ['form.eligibility.result-note.deadline',
-                                'form.eligibility.result-note.capital_investment']
+        expected_information = [_('form.eligibility.result-note.capital_investment'),
+                                _('form.eligibility.result-note.deadline')]
```
Comment (frederike-ramin) [inline webapp/tests/forms/steps/test_eligibility_steps.py:3170]: I don't think this should be necessary as the `_` function is mocked for the function call beneath?

File: webapp/tests/forms/steps/test_eligibility_steps.py [hunk 5]
```diff
@@ -3274,3 +3272,138 @@ def test_if_no_user_b_elster_account_and_no_cheaper_check_then_set_no_info(self)
...
+        with patch('app.forms.steps.eligibility_steps._', MagicMock(side_effect=lambda text_id: text_id)):
+            step = EligibilityMaybeDisplaySteuerlotseStep(
+                endpoint='eligibility',
+                stored_data=session_data,
+                render_info=EligibilityMaybeDisplaySteuerlotseStep.prepare_render_info(
+                    {})
+            )
+            step.handle()
+
+        self.assertEqual(expected_information, step.render_info.additional_info['dependent_notes'])
```
Comment (frederike-ramin) [inline webapp/tests/forms/steps/test_eligibility_steps.py:3301]: This should actually never happen for the MaybeStep,  I think. If person B has no elster account, it will always be the success page that is shown, right?

Comment (frederike-ramin) [reply webapp/tests/forms/steps/test_eligibility_steps.py:3301]: I guess the test we want here is `test_if_both_users_have_elster_account_then_set_correct_info`

File: webapp/tests/forms/steps/test_eligibility_steps.py [hunk 6]
```diff
@@ -3274,3 +3272,138 @@ def test_if_no_user_b_elster_account_and_no_cheaper_check_then_set_no_info(self)
...
+
+    def test_if_user_b_has_no_elster_account_and_user_has_minimal_investment_income_check_then_set_correct_info(self):
```
Comment (frederike-ramin) [inline webapp/tests/forms/steps/test_eligibility_steps.py:3374]: Also with those two tests, the condition should be both users.

File: webapp/tests/forms/steps/test_eligibility_steps.py [hunk 7]
```diff
@@ -3274,3 +3272,138 @@ def test_if_no_user_b_elster_account_and_no_cheaper_check_then_set_no_info(self)
...
+    def test_if_no_user_b_elster_account_and_no_cheaper_check_then_set_no_info(self):
+        expected_information = [_('form.eligibility.result-note.deadline')]
+        with patch('app.forms.steps.eligibility_steps._', MagicMock(side_effect=lambda text_id: text_id)):
+            step = EligibilityMaybeDisplaySteuerlotseStep(
+                    endpoint='eligibility',
+                    render_info=EligibilityMaybeDisplaySteuerlotseStep.prepare_render_info(
+                        {}))
+            step.handle()
+
+        self.assertEqual(expected_information, step.render_info.additional_info['dependent_notes'])
```
Comment (frederike-ramin) [inline webapp/tests/forms/steps/test_eligibility_steps.py:3409]: Can we maybe rename the test? As it doesn't test for "no info"? I'd suggest `test_if_no_special_case_then_set_only_deadline_info`

**Review by frederike-ramin — CHANGES_REQUESTED** [review]
I think there is really just one mistake in the way the notes are shown. It's easily fixable but the changing of tests could be annoying. Let me know if you'd like support for that. In fact, I might be wrong about the conditions in general, then let's chat about it. :)

File: webapp/app/model/eligibility_data.py (hunk 1 above)
Comment (SannyNguyenHung) [reply webapp/app/model/eligibility_data.py:478]: Change it back! Thought it should be last, but since there are no other previous steps, we should leave it on top.

File: webapp/app/translations/de/LC_MESSAGES/messages.po (hunk 3 above)
Comment (SannyNguyenHung) [reply webapp/app/translations/de/LC_MESSAGES/messages.po:1]: Sry!

File: webapp/tests/forms/steps/test_eligibility_steps.py (hunk 4 above)
Comment (SannyNguyenHung) [reply webapp/tests/forms/steps/test_eligibility_steps.py:3170]: Mocked the _l functions

File: webapp/app/forms/steps/eligibility_steps.py (hunk 2 above)
Comment (SannyNguyenHung) [reply webapp/app/forms/steps/eligibility_steps.py:778]: Oh thanks for the hint! :sweat_smile:

File: webapp/tests/forms/steps/test_eligibility_steps.py (hunk 5 above)
Comment (SannyNguyenHung) [reply webapp/tests/forms/steps/test_eligibility_steps.py:3301]: Thanks!

**Comment by SannyNguyenHung** [comment]
Thank you @frederike-ramin :smiley:

**Review by frederike-ramin — APPROVED** [review]
Thank you! So nice now^^
